### PR TITLE
Java: Cleanup imports of `ExternalFlow`

### PR DESCRIPTION
**Pull request by michaelnebel**
In this PR we cleanup the bi-directional *framework* imports that were introduced for Models as Data.
The bi-directional imports created a huge SCC of dependencies and removing the importa revealed some missing imports in other parts of the code that were relying on the framework SCC in `ExternalFlow.qll`.
The code has been modified accordingly and bi-directional imports because of other abstract classes have been introduced where needed.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/frameworks/Jdbc.qll [hunk 1]
```diff
@@ -2,7 +2,7 @@
  * Provides classes and predicates for working with the Java JDBC API.
  */
 
-private import semmle.code.java.dataflow.ExternalFlow
+private import java
```
Comment (atorralba) [inline java/ql/lib/semmle/code/java/frameworks/Jdbc.qll:5]: NIT: I guess these `private import java` don't need to be `private` (although it doesn't hurt...)

File: java/ql/lib/semmle/code/java/dataflow/FlowSources.qll [hunk 2]
```diff
@@ -36,6 +36,14 @@ abstract class RemoteFlowSource extends DataFlow::Node {
   abstract string getSourceType();
 }
 
+/**
+ * A module for importing frameworks that define remote flow sources.
+ */
+private module RemoteFlowSources {
+  private import semmle.code.java.frameworks.android.Widget
+  private import semmle.code.java.security.TemplateInjection
```
Comment (atorralba) [inline java/ql/lib/semmle/code/java/dataflow/FlowSources.qll:44]: I don't think `TemplateInjection` defines new remote sources. It only has:

https://github.com/github/codeql/blob/ef9b503f69bb7e9f7f5d91a7dce2520d8b3f6380/java/ql/lib/semmle/code/java/security/TemplateInjection.qll#L65

Which is a remote flow source _use_, not a new definition.

```suggestion
```

File: java/ql/lib/semmle/code/java/frameworks/Regex.qll [hunk 3]
```diff
@@ -1,6 +1,6 @@
 /** Definitions related to `java.util.regex`. */
 
-private import semmle.code.java.dataflow.ExternalFlow
+private import java
```
Comment (atorralba) [inline java/ql/lib/semmle/code/java/frameworks/Regex.qll:3]: Same thing here.

**Review by atorralba — COMMENTED** [review]
Mostly looks good to me, thanks for doing this! I added a couple NITs (that you can completely ignore), and one minor fix.

File: java/ql/lib/semmle/code/java/dataflow/FlowSources.qll (hunk 2 above)
Comment (michaelnebel) [reply java/ql/lib/semmle/code/java/dataflow/FlowSources.qll:44]: Oh - you are right! Good catch! Will fix.
I just searched for regexp `extends (.*)RemoteFlowSource` and did not consider this case (and didn't pay attention to the actual source code).

File: java/ql/lib/semmle/code/java/frameworks/Jdbc.qll (hunk 1 above)
Comment (michaelnebel) [reply java/ql/lib/semmle/code/java/frameworks/Jdbc.qll:5]: Yes, I just didn't want to re-expose something that was not exposed before as this is a library, but I will change it for consistency.

**Review by atorralba — APPROVED** [review]

**Comment by michaelnebel** [comment]
> Wondering if this is worth a DCA run 🤔 (when it works).

This is definitely worth a DCA run! We need to see if there are any sudden drops in performance (or improvements).
We could risk worse performance, if some bi-directional imports have been overlooked that could cause re-evaluation.
Will not merge before we are able to run DCA.

**Comment by michaelnebel** [comment]
DCA looks good.
- The overall performance looks acceptable.
- There appears to be some minor differences in a couple of the query execution times (stages), but an overall reduction in the execution time by all stages. This might be a glitch or maybe a change in the execution order.

Merging now.

**Pull request merged**